packages/mui-material/src/styles/createPalette.ts

```diff
@@ -83,13 +224,18 @@ function getDark() {
 
 export const dark = getDark();
 
-function addLightOrDark(intent, direction, shade, tonalOffset) {
-  const tonalOffsetLight = tonalOffset.light || tonalOffset;
-  const tonalOffsetDark = tonalOffset.dark || tonalOffset * 1.5;
+function addLightOrDark(
+  intent: PaletteColor,
+  direction: 'light' | 'dark',
+  shade: number | string,
+  tonalOffset: PaletteTonalOffset,
+) {
+  const tonalOffsetLight = typeof tonalOffset === 'number' ? tonalOffset : tonalOffset.light;
+  const tonalOffsetDark = typeof tonalOffset === 'number' ? tonalOffset * 1.5 : tonalOffset.dark;
 
   if (!intent[direction]) {
     if (intent.hasOwnProperty(shade)) {
-      intent[direction] = intent[shade];
+      intent[direction] = intent[shade as keyof typeof intent] as string;
```

Comment:
We can directly use:
```suggestion
      intent[direction] = intent[shade as keyof PaletteColor] as string;
```
since we know the shape at compile time. Same in other places.